---
title: "AbSeq"
header-evals:
  - \usepackage{comment}
output: 
  flexdashboard::flex_dashboard:
    vertical_layout: scroll
params:
    # img dimensions
    width: 8
    height: 5
    # ggplotly?
    interactive: TRUE
    # eval D-gene abundance related plots? - light chain might not want this
    inclD: TRUE
    # sample name
    name: "PLACEHOLDER SAMPLE NAME"
    # single sample?
    single: TRUE
    # root output directory
    rootDir: "/newhome/jiahong/sandbox/immulator_out/output_results/report/PCR1_ACGT_L001_crossed"
    hasAnnot: TRUE
    hasAbun: TRUE
---

<style type="text/css">
.chart-title {  /* increase chart title font size */
   font-size: 24px;
</style>

```{r setup, include=FALSE}
library(flexdashboard)
library(plotly)
library(png)
library(grid)

.getFile <- function(path, pattern) {
    flist <- list.files(path = path, pattern = pattern, full.names = T)
    n <- length(flist)
    if (n != 1) {
        stop(paste("Expected to find one file, but found", n, "from",
                   paste(flist, collapse = ", "), "instead"))
    }
    return(flist[[1]])
}
```

\begin{comment}
Page 1 is a overview of the sample analysis
\end{comment}

`r if(!params$hasAnnot) {"\\begin{comment}"}`

Summary {data-orientation=rows}
=======================================================================

Row
-----------------------------------------------------------------------

### Sequence lengths without outlier removal

```{r, eval=params$hasAnnot, fig.width=params$width, fig.height=params$height}
fname <- .getFile(path = file.path(params$rootDir, "annot"),
                  pattern = ".*_all_clones_len_dist\\.Rdata")
load(fname)
if (params$interactive) {
    ggplotly(plot)
} else {
    plot
}
```

Row
-----------------------------------------------------------------------

### Sequence lengths __with__ outlier removal

```{r, eval=params$hasAnnot, fig.width=params$width, fig.height=params$height}
fname <- .getFile(path = file.path(params$rootDir, "annot"),
                  pattern = ".*_all_clones_len_dist_no_outliers\\.Rdata")
load(fname)
if (params$interactive) {
    ggplotly(plot)
} else {
    plot
}
```

`r if(!params$hasAnnot) {"\\end{comment}"}`

\begin{comment}
Page 2 is abundance analysis
\end{comment}

`r if(!params$hasAbun) {"\\begin{comment}"}`

Abundance {data-orientation=rows}
=======================================================================


Row
-----------------------------------------------------------------------

### V family 

```{r, eval=params$hasAbun, fig.width=params$width, fig.height=params$height}
fname <- .getFile(path = file.path(params$rootDir, "abundance"),
                  pattern = ".*igv_dist_family_level\\.Rdata")
load(fname)
if (params$interactive) {
    ggplotly(plot)
} else {
    plot
}
```

### V gene

```{r, eval=params$hasAbun, fig.width=params$height, fig.height=params$width}
fname <- .getFile(path = file.path(params$rootDir, "abundance"),
                  pattern = ".*igv_dist_gene_level\\.Rdata")
load(fname)
# TODO: ggplotly is buggy with horizontal plot in tabsets
if (params$interactive) {
    plot
    #ggplotly(plot)
} else {
    plot
}
```

> Only the top 15 from each sample are shown

`r if(!params$inclD || !params$hasAbun) {"\\begin{comment}"}`

Row
-----------------------------------------------------------------------

### D family

```{r, eval=(params$inclD && params$hasAbun), fig.width=params$width, fig.height=params$height}
fname <- .getFile(path = file.path(params$rootDir, "abundance"),
                  pattern = ".*igd_dist_family_level\\.Rdata")
load(fname)
if (params$interactive) {
    ggplotly(plot)
} else {
    plot
}
```

### D gene

```{r, eval=(params$inclD && params$hasAbun), fig.width=params$height, fig.height=params$width}
fname <- .getFile(path = file.path(params$rootDir, "abundance"),
                  pattern = ".*igd_dist_gene_level\\.Rdata")
load(fname)
# TODO: ggplotly is buggy with horizontal plot in tabsets
if (params$interactive) {
    plot
    #ggplotly(plot)
} else {
    plot
}
```

> Only the top 15 from each sample are shown

`r if(!params$inclD || !params$hasAbun) {"\\end{comment}"}`


Row
-----------------------------------------------------------------------

### J family

```{r, eval=params$hasAbun, fig.width=params$width, fig.height=params$height}
fname <- .getFile(path = file.path(params$rootDir, "abundance"),
                  pattern = ".*igj_dist_family_level\\.Rdata")
load(fname)
if (params$interactive) {
    ggplotly(plot)
} else {
    plot
}
```

### J variant

```{r, eval=params$hasAbun, fig.width=params$height, fig.height=params$width}
fname <- .getFile(path = file.path(params$rootDir, "abundance"),
                  pattern = ".*igj_dist_variant_level\\.Rdata")
load(fname)
# TODO: ggplotly is buggy with horizontal plot in tabsets
if (params$interactive) {
    plot
    #ggplotly(plot)
} else {
    plot
}
```

> Only the top 15 from each sample are shown

`r if(!params$single || !params$hasAbun) {"\\begin{comment}"}`

Row
-----------------------------------------------------------------------

### V-J association

```{r, eval=(params$single && params$hasAbun), fig.width=params$width, fig.height=params$height}
fname <- .getFile(path = file.path(params$rootDir, "abundance"), pattern = ".*_vjassoc\\.png")
img <- readPNG(fname)
grid.raster(img)
```
`r if(!params$single || !params$hasAbun) {"\\end{comment}"}`

> Plot associations between V genes and J genes on the family level

`r if(!params$hasAbun) {"\\end{comment}"}`


Row {.tabset .tabset-fade}
-----------------------------------------------------------------------

### Bitscore vs Alignment Length
```{r, eval=params$hasAbun, fig.width=params$width, fig.height=params$height}
fname <- .getFile(path = file.path(params$rootDir, "abundance"),
                  pattern = ".*_igv_align_quality_bitscore_hm\\.png")
img <- readPNG(fname)
grid.raster(img)
```

### Gaps vs Alignment Length
```{r, eval=params$hasAbun, fig.width=params$width, fig.height=params$height}
fname <- .getFile(path = file.path(params$rootDir, "abundance"),
                  pattern = ".*_igv_align_quality_gaps_hm\\.png")
img <- readPNG(fname)
grid.raster(img)
```

### Identity vs Alignment Length
```{r, eval=params$hasAbun, fig.width=params$width, fig.height=params$height}
fname <- .getFile(path = file.path(params$rootDir, "abundance"),
                  pattern = ".*_igv_align_quality_identity_hm\\.png")
img <- readPNG(fname)
grid.raster(img)
```

### Mismatches vs Alignment Length
```{r, eval=params$hasAbun, fig.width=params$width, fig.height=params$height}
fname <- .getFile(path = file.path(params$rootDir, "abundance"),
                  pattern = ".*_igv_align_quality_mismatches_hm\\.png")
img <- readPNG(fname)
grid.raster(img)
```

### Subject Start vs Query Start
```{r, eval=params$hasAbun, fig.width=params$width, fig.height=params$height}
fname <- .getFile(path = file.path(params$rootDir, "abundance"),
                  pattern = ".*_igv_align_quality_start_hm\\.png")
img <- readPNG(fname)
grid.raster(img)
```

Row
-----------------------------------------------------------------------

### Indels (Gaps)
```{r, eval=params$hasAbun, fig.width=params$width, fig.height=params$height}
fname <- .getFile(path = file.path(params$rootDir, "abundance"),
                  pattern = ".*_igv_gaps_dist\\.Rdata")
load(fname)
if (params$interactive) {
    ggplotly(plot)
} else {
    plot
}
```

> Plot of insertions and deletions found within the __V gene__

### Mismatches
```{r, eval=params$hasAbun, fig.width=params$width, fig.height=params$height}
fname <- .getFile(path = file.path(params$rootDir, "abundance"),
                  pattern = ".*_igv_mismatches_dist\\.Rdata")
load(fname)
if (params$interactive) {
    ggplotly(plot)
} else {
    plot
}
```

> Plot of nucleotide base mismatches found within the __V gene__

\begin{comment}
=======================================================================
-----------------------------------------------------------------------
\end{comment}
