---
title: "AbSeq"
header-evals:
  - \usepackage{comment}
output: 
  flexdashboard::flex_dashboard:
    vertical_layout: scroll
params:
    # img dimensions
    width: 8
    height: 5
    # ggplotly?
    interactive: TRUE
    # eval D-gene abundance related plots? - light chain might not want this
    inclD: TRUE
    # sample name
    name: "PLACEHOLDER SAMPLE NAME"
    # single sample?
    single: FALSE
    # root output directory
    rootDir: "/newhome/jiahong/sandbox/immulator_out/output_results/report/PCR1_ACGT_L001_crossed_vs_PCR2_ACGT_L001_crossed"
    hasAnnot: TRUE
    hasAbun: TRUE
    hasProd: TRUE
    hasDiv: TRUE
---

<style type="text/css">
.chart-title {  /* increase chart title font size */
   font-size: 24px;
}
</style>

```{r setup, include=FALSE}
library(flexdashboard)
library(plotly)
library(png)
library(grid)
library(gridExtra)

.getFile <- function(path, pattern, .stop = F) {
    flist <- list.files(path = path, pattern = pattern, full.names = T)
    n <- length(flist)
    if (n != 1) {
        if (.stop) {
            stop(paste("Expected to find one file, but found", n, "from",
                       paste(flist, collapse = ", "), "instead"))
        } else {
            return(NA)
        }
    }
    return(flist[[1]])
}
```

<!-- 
Page 1 is a overview of the sample analysis
-->

`r if(!params$hasAnnot) {"\\begin{comment}"}`

Summary {data-orientation=rows}
=======================================================================

Row
-----------------------------------------------------------------------

### Sequence lengths without outlier removal

```{r, eval=params$hasAnnot, fig.width=params$width, fig.height=params$height}
fname <- .getFile(path = file.path(params$rootDir, "annot"),
                  pattern = ".*_all_clones_len_dist\\.Rdata")
if (!is.na(fname)) {
    load(fname)
    if (params$interactive) {
        ggplotly(plot)
    } else {
        plot
    }
} else {
    cat("404!")
}
```

Row
-----------------------------------------------------------------------

### Sequence lengths __with__ outlier removal

```{r, eval=params$hasAnnot, fig.width=params$width, fig.height=params$height}
fname <- .getFile(path = file.path(params$rootDir, "annot"),
                  pattern = ".*_all_clones_len_dist_no_outliers\\.Rdata")
if (!is.na(fname)) {
    load(fname)
    if (params$interactive) {
        ggplotly(plot)
    } else {
        plot
    }
} else {
    cat("404!")
}
```

`r if(!params$hasAnnot) {"\\end{comment}"}`

<!-- Page 2 is abundance analysis -->

`r if(!params$hasAbun) {"\\begin{comment}"}`

Abundance {data-orientation=rows}
=======================================================================


Row
-----------------------------------------------------------------------

### V family 

```{r, eval=params$hasAbun, fig.width=params$width, fig.height=params$height}
fname <- .getFile(path = file.path(params$rootDir, "abundance"),
                  pattern = ".*igv_dist_family_level\\.Rdata")
if (!is.na(fname)) {
    load(fname)
    if (params$interactive) {
        ggplotly(plot)
    } else {
        plot
    }
} else {
    cat("404!")
}
```

### V gene

```{r, eval=params$hasAbun, fig.width=params$height, fig.height=params$width}
fname <- .getFile(path = file.path(params$rootDir, "abundance"),
                  pattern = ".*igv_dist_gene_level\\.Rdata")
if (!is.na(fname)) {
    load(fname)
    # TODO: ggplotly is buggy with horizontal plot in tabsets
    if (params$interactive) {
        plot
        #ggplotly(plot)
    } else {
        plot
    }
} else {
    cat("404!")
}
```

> Only the top 15 from each sample are shown

`r if(!params$inclD) {"\\begin{comment}"}`

Row
-----------------------------------------------------------------------

### D family

```{r, eval=(params$inclD && params$hasAbun), fig.width=params$width, fig.height=params$height}
fname <- .getFile(path = file.path(params$rootDir, "abundance"),
                  pattern = ".*igd_dist_family_level\\.Rdata")
if (!is.na(fname)) {
    load(fname)
    if (params$interactive) {
        ggplotly(plot)
    } else {
        plot
    }
} else {
    cat("404!")
}
```

### D gene

```{r, eval=(params$inclD && params$hasAbun), fig.width=params$height, fig.height=params$width}
fname <- .getFile(path = file.path(params$rootDir, "abundance"),
                  pattern = ".*igd_dist_gene_level\\.Rdata")
if (!is.na(fname)) {
    load(fname)
    # TODO: ggplotly is buggy with horizontal plot in tabsets
    if (params$interactive) {
        plot
        #ggplotly(plot)
    } else {
        plot
    }
} else {
    cat("404!")
}
    
```

> Only the top 15 from each sample are shown

`r if(!params$inclD) {"\\end{comment}"}`


Row
-----------------------------------------------------------------------

### J family

```{r, eval=params$hasAbun, fig.width=params$width, fig.height=params$height}
fname <- .getFile(path = file.path(params$rootDir, "abundance"),
                  pattern = ".*igj_dist_family_level\\.Rdata")
if (!is.na(fname)) {
    load(fname)
    if (params$interactive) {
        ggplotly(plot)
    } else {
        plot
    }
} else {
    cat("404!")
}
```

### J variant

```{r, eval=params$hasAbun, fig.width=params$height, fig.height=params$width}
fname <- .getFile(path = file.path(params$rootDir, "abundance"),
                  pattern = ".*igj_dist_variant_level\\.Rdata")
if (!is.na(fname)) {
    load(fname)
    # TODO: ggplotly is buggy with horizontal plot in tabsets
    if (params$interactive) {
        plot
        #ggplotly(plot)
    } else {
        plot
    }
} else {
    cat("404!")
}
```

> Only the top 15 from each sample are shown

`r if(!params$single) {"\\begin{comment}"}`

Row
-----------------------------------------------------------------------

### V-J association

```{r, eval=(params$single && params$hasAbun), fig.width=params$width, fig.height=params$height}
fname <- .getFile(path = file.path(params$rootDir, "abundance"), pattern = ".*_vjassoc\\.png")
if (!is.na(fname)) {
    img <- readPNG(fname)
    grid.raster(img)
} else {
    cat("404!")
}
```

> Plot associations between V genes and J genes on the family level




Row {.tabset .tabset-fade}
-----------------------------------------------------------------------
```{r, echo = FALSE, eval = params$single}
alignQual <- list("Bitscore vs Alignment Length" = "bitscore",
               "Gaps vs Alignment Length" = "gaps",
               "Identity vs Alignment Length" = "identity",
               "Mismatches vs Alignment Length" = "mismatches",
               "Subject Start vs Query Start" = "start")
alignmentQuality <- lapply(seq_along(alignQual), function(i) {
    line1 <- knitr::knit_expand(text = sprintf("### %s\n", names(alignQual)[i]))
    line2 <- knitr::knit_expand(text = sprintf("\n```{r align_qual_chunk_%d, eval=params$hasAbun, fig.width=params$width, fig.height=params$height}", i))
    line3 <- knitr::knit_expand(text = sprintf('\nfname <- .getFile(path = file.path(params$rootDir, "abundance"), pattern = ".*_igv_align_quality_%s_hm\\\\.png")', alignQual[[i]]))
    line4 <- knitr::knit_expand(text = '\nif(!is.na(fname)) { img <- readPNG(fname); grid.raster(img) } else { cat("404!") }')
    line5 <- knitr::knit_expand(text = "\n```\n")
    paste(line1, line2, line3, line4, line5, collapse = "\n")
})
```

`r if (params$single) { paste(knitr::knit(text = paste(alignmentQuality, collapse = "\n"))) }`

`r if(!params$single) {"\\end{comment}"}`

Row
-----------------------------------------------------------------------

### Indels (Gaps)
```{r, eval=params$hasAbun, fig.width=params$width, fig.height=params$height}
fname <- .getFile(path = file.path(params$rootDir, "abundance"),
                  pattern = ".*_igv_gaps_dist\\.Rdata")
if (!is.na(fname)) {
    load(fname)
    if (params$interactive) {
        ggplotly(plot)
    } else {
        plot
    }
} else {
    cat("404!")
}
```

> Plot of insertions and deletions found within the __V gene__

### Mismatches
```{r, eval=params$hasAbun, fig.width=params$width, fig.height=params$height}
fname <- .getFile(path = file.path(params$rootDir, "abundance"),
                  pattern = ".*_igv_mismatches_dist\\.Rdata")
if (!is.na(fname)) {
    load(fname)
    if (params$interactive) {
        ggplotly(plot)
    } else {
        plot
    }
} else {
    cat("404!")
}
```

> Plot of nucleotide base mismatches found within the __V gene__

`r if(!params$hasAbun) {"\\end{comment}"}`


`r if(!params$hasProd) {"\\start{comment}"}`

Productivity {data-orientation=rows}
=======================================================================

Row
-----------------------------------------------------------------------

### Productivity summary

```{r, eval=params$hasProd, fig.width=params$width, fig.height=params$height}
fname <- .getFile(path = file.path(params$rootDir, "productivity"),
                  pattern = ".*_productivity\\.Rdata")
if (!is.na(fname)) {
    load(fname)
    if (params$interactive) {
        # TODO: buggy facet plot
        plot
        # ggplotly(plot)
    } else {
        plot
    }
} else {
    cat("404!")
}
```

> Productivity distribution of sample based on stop codons and frameshifts

### Frameshift distribution

```{r, eval=params$hasProd, fig.width=params$width, fig.height=params$height}
fname <- .getFile(path = file.path(params$rootDir, "productivity"),
                  pattern = ".*_crossed_vjframe_dist\\.Rdata")

if (!is.na(fname)) {
    load(fname)
    if (params$interactive) {
        ggplotly(plot)
    } else {
        plot
    }
} else {
    cat("404!")
}
```

> Distribution of in-frame and out-of-frame sequences

Row {.tabset .tabset-fade}
-----------------------------------------------------------------------
<!--  Example of how it should look like after knitting
### CDR1
fname <- .getFile(path = file.path(params$rootDir, "productivity"),
                  pattern = ".*_cdr1_mismatches_dist\\.Rdata")

if (!is.na(fname)) {
    load(fname)
    if (params$interactive) {
        mismatch <- ggplotly(plot + labs(title = "Mismatches, gaps and gaps in out-of-frame sequences in CDR1"))
    } else {
        mismatch <- plot
    }
} else {
    df <- data.frame()      # give user empty graph
    mismatch  <- ggplot(df), + geom_point() + xlim(0, 10) + ylim(0, 100)
}

fname <- .getFile(path = file.path(params$rootDir, "productivity"),
                  pattern = ".*_cdr1_gaps_dist\\.Rdata")
if (!is.na(fname)) {
    load(fname)
    if (params$interactive) {
        gaps <- ggplotly(plot + labs(title = ""))
    } else {
        gaps <- plot
    }
} else {
    df <- data.frame()      # give user empty graph
    gaps <- ggplot(df), + geom_point() + xlim(0, 10) + ylim(0, 100)
}

fname <- .getFile(path = file.path(params$rootDir, "productivity"),
                  pattern = ".*_cdr1_gaps_dist_out_of_frame\\.Rdata")
if (!is.na(fname)) {
    load(fname)
    if (params$interactive) {
        gapsOut <- ggplotly(plot + labs(title = ""))
    } else {
        gapsOut <- plot
    }
} else {
    df <- data.frame()      # give user empty graph
    gapsOut  <- ggplot(df), + geom_point() + xlim(0, 10) + ylim(0, 100)
}

if (params$interactive) {
    subplot(mismatch, gaps, gapsOut, titleX = TRUE, titleY = TRUE)
} else {
    grid.arrange(mismatch, gaps, gapsOut, nrow = 1)
}
> Mismatches, gaps, gaps (in out-of-frame sequences) distribution in %s region\n
-->

```{r echo = FALSE, eval = TRUE}
titles <- c("CDR1", "CDR2", "CDR3", "FR1", "FR2", "FR3")
frcdrMisGaps <- lapply(seq_along(titles), function(i) {
    line1 <- knitr::knit_expand(text = sprintf("### %s\n", titles[i]))
    line2 <- knitr::knit_expand(text = sprintf("\n```{r frcdr_gap_mis_chunk_%d, eval=params$hasProd, fig.width=params$width*3, fig.height=params$height}", i))
    line3 <- knitr::knit_expand(text = sprintf('\nfname <- .getFile(file.path(params$rootDir, "productivity"), pattern = ".*_%s_mismatches_dist\\\\.Rdata")', tolower(titles[i])))
    line4 <- knitr::knit_expand(text = sprintf('\nif (!is.na(fname)) { load(fname); if (params$interactive) { mismatch <- ggplotly(plot + labs(title = "Mismatches, gaps and gaps in out-of-frame sequences in %s")) } else { mismatch <- plot } } else { df <- data.frame(); mismatch <- ggplot(df) + geom_point() + xlim(0, 10) + ylim(0, 100) }', titles[i]))
    line5 <- knitr::knit_expand(text = sprintf('\nfname <- .getFile(file.path(params$rootDir, "productivity"), pattern = ".*_%s_gaps_dist\\\\.Rdata")', tolower(titles[i])))
    line6 <- knitr::knit_expand(text = '\nif (!is.na(fname)) { load(fname); if (params$interactive) { gaps <- ggplotly(plot + labs(title = "")) } else { gaps <- plot } } else { df <- data.frame(); gaps <- ggplot(df) + geom_point() + xlim(0, 10) + ylim(0, 100) }')
    line7 <- knitr::knit_expand(text = sprintf('\nfname <- .getFile(file.path(params$rootDir, "productivity"), pattern = ".*_%s_gaps_dist_out_of_frame\\\\.Rdata")', tolower(titles[i])))
    line8 <- knitr::knit_expand(text = '\nif (!is.na(fname)) { load(fname); if (params$interactive) { gapsOut <- ggplotly(plot + labs(title = "")) } else { gapsOut <- plot } } else { df <- data.frame(); gapsOut <- ggplot(df) + geom_point() + xlim(0, 10) + ylim(0, 100) }')
    line9 <- knitr::knit_expand(text = "\nif (params$interactive) { subplot(mismatch, gaps, gapsOut, titleX = T, titleY = T) } else { grid.arrange(mismatch, gaps, gapsOut, nrow = 1)}")
    line10 <- knitr::knit_expand(text = "\n```\n")
    line11 <- knitr::knit_expand(text = sprintf("\n> Mismatches, gaps and gaps (in out-of-frame sequences) distribution in %s region\n", titles[i]))
    paste(line1, line2, line3, line4, line5, line6, line7, line8, line9, line10,
          line11, collapse = "\n")
})
```

`r paste(knitr::knit(text = paste(frcdrMisGaps, collapse = "\n")))`

<!-- Example of this knitr output:
### IGV
fname <- .getFile(path = file.path(params$rootDir, "productivity"),
                  pattern = ".*_igv_mismatches_dist\\.Rdata")
# take the titles away if it's interactive because subplots will only use the first
if (!is.na(fname)) {
    load(fname)
    if (params$interactive) {
        mismatch <- ggplotly(plot + labs(title = "Mismatches and gaps in IGV"))
    } else {
        mismatch <- plot
    }
} else {
    df <- data.frame()      # give user empty graph
    mismatch <- ggplot(df), + geom_point() + xlim(0, 10) + ylim(0, 100)
}

fname <- .getFile(path = file.path(params$rootDir, "productivity"),
                  pattern = ".*_igv_gaps_dist\\.Rdata")
if (is.na(fname)) {}
    load(fname)
    if (params$interactive) {
        gaps <- ggplotly(plot + labs(title = ""))
    } else {
        gaps <- plot
    }
} else {
    df <- data.frame()      # give user empty graph
    gaps <- ggplot(df), + geom_point() + xlim(0, 10) + ylim(0, 100)
}

if (params$interactive) {
    subplot(mismatch, gaps, titleX = TRUE, titleY = TRUE)
} else {
    grid.arrange(mismatch, gaps, nrow = 1)
}
> Mismatches and gaps distribution in V gene
-->

```{r echo = FALSE, eval = TRUE}
if (params$inclD) {
    titles <- c("IGV", "IGD", "IGJ")
} else {
    titles <- c("IGV", "IGJ")
}
vdjMisGaps <- lapply(seq_along(titles), function(i) {
    line1 <- knitr::knit_expand(text = sprintf("### %s\n", titles[i]))
    line2 <- knitr::knit_expand(text = sprintf("\n```{r vdj_gap_mis_chunk_%d, eval=params$hasProd, fig.width=params$width*3, fig.height=params$height}", i))
    line3 <- knitr::knit_expand(text = sprintf('\nfname <- .getFile(file.path(params$rootDir, "productivity"), pattern = ".*_%s_mismatches_dist\\\\.Rdata")', tolower(titles[i])))
    line4 <- knitr::knit_expand(text = sprintf('\nif (!is.na(fname)) { load(fname); if (params$interactive) { mismatch <- ggplotly(plot + labs(title = "Mismatches and gaps in %s")) } else { mismatch <- plot } } else { df <- data.frame(); mismatch <- ggplot(df) + geom_point() + xlim(0, 10) + ylim(0, 100) }', titles[i]))
    line5 <- knitr::knit_expand(text = sprintf('\nfname <- .getFile(file.path(params$rootDir, "productivity"), pattern = ".*_%s_gaps_dist\\\\.Rdata")', tolower(titles[i])))
    line6 <- knitr::knit_expand(text = '\nif (!is.na(fname)) { load(fname); if (params$interactive) { gaps <- ggplotly(plot + labs(title = "")) } else { gaps <- plot } } else { df <- data.frame(); gaps <- ggplot(df) + geom_point() + xlim(0, 10) + ylim(0, 100) }')
    line7 <- knitr::knit_expand(text = "\nif (params$interactive) { subplot(mismatch, gaps, titleX = T, titleY = T) } else { grid.arrange(mismatch, gaps, nrow = 1)}")
    line8 <- knitr::knit_expand(text = "\n```\n")
    line9 <- knitr::knit_expand(text = sprintf("\n> Mismatchesa and gaps distribution in %s\n", titles[i]))
    paste(line1, line2, line3, line4, line5, line6, line7, line8, line9, collapse = "\n")
})
```

`r paste(knitr::knit(text = paste(vdjMisGaps, collapse = "\n")))`


Row {.tabset .tabset-fade}
-----------------------------------------------------------------------

### IGV productive distribution

```{r, eval=params$hasProd, fig.width=params$width, fig.height=params$height}
fname <- .getFile(path = file.path(params$rootDir, "productivity"),
                  pattern = ".*_igv_dist_productive\\.Rdata")

if (!is.na(fname)) {
    load(fname)
    if (params$interactive) {
        ggplotly(plot)
    } else {
        plot
    }
} else {
    cat("404!")
}
```

> Distribution of productive sequences from IGV family

### IGV inframe, unproductive distribution

```{r, eval=params$hasProd, fig.width=params$width, fig.height=params$height}
fname <- .getFile(path = file.path(params$rootDir, "productivity"),
                  pattern = ".*_igv_dist_inframe_unproductive\\.Rdata")

if (!is.na(fname)) {
    load(fname)
    if (params$interactive) {
        ggplotly(plot)
    } else {
        plot
    }
} else {
    cat("404!")
}
```

> Distribution of in-frame but unproductive sequences from IGV family

### IGV out-of-frame distribution

```{r, eval=params$hasProd, fig.width=params$width, fig.height=params$height}
fname <- .getFile(path = file.path(params$rootDir, "productivity"),
                  pattern = ".*_igv_dist_out_of_frame\\.Rdata")

if (!is.na(fname)) {
    load(fname)
    if (params$interactive) {
        ggplotly(plot)
    } else {
        plot
    }
} else {
    cat("404!")
}
```

> Distribution of out-of-frame sequences from IGV family


Row {.tabset .tabset-fade}
-----------------------------------------------------------------------

### Inframe

```{r, eval=params$hasProd, fig.width=params$width, fig.height=params$height}
fname <- .getFile(path = file.path(params$rootDir, "productivity"),
                  pattern = ".*_stopcodon_region_inframe\\.Rdata", .stop = F)
if (is.na(fname)) {
    cat(paste("Plot is not available for this sample, probably because there's",
                "no such distribution present in this sample"))
} else {
    load(fname)
    if (params$interactive) {
        ggplotly(plot)
    } else {
        plot
    }
}

```

> Distribution of stop codons in CDR/FR regions from in-frame sequences

### Out-of-frame
```{r, eval=params$hasProd, fig.width=params$width, fig.height=params$height}
fname <- .getFile(path = file.path(params$rootDir, "productivity"),
                  pattern = ".*_stopcodon_region_outframe\\.Rdata", .stop = F)
if (is.na(fname)) {
    cat(paste("Plot is not available for this sample, probably because there's",
                "no such distribution present in this sample"))
} else {
    load(fname)
    if (params$interactive) {
        ggplotly(plot)
    } else {
        plot
    }
}
```

> Distribution of stop codons in CDR/FR regions from out-of-frame sequences


<!--
=======================================================================
-----------------------------------------------------------------------
-->
