---
title: "AbSeq"
header-evals:
  - \usepackage{comment}
output: 
  flexdashboard::flex_dashboard:
    vertical_layout: scroll
params:
    # img dimensions
    width: 8
    height: 5
    # ggplotly?
    interactive: TRUE
    # eval D-gene abundance related plots? - light chain might not want this
    inclD: TRUE
    # sample name
    name: "PLACEHOLDER SAMPLE NAME"
    # single sample?
    single: TRUE
    # root output directory
    rootDir: "/newhome/jiahong/sandbox/immulator_out/output_results/report/PCR1_ACGT_L001_crossed"
    hasAnnot: FALSE
    hasAbun: FALSE
    hasProd: FALSE
    hasDiv: TRUE
---

<style type="text/css">
.chart-title {  /* increase chart title font size */
   font-size: 24px;
}
</style>

```{r setup, include=FALSE}
library(flexdashboard)
library(plotly)
library(png)
library(grid)
library(gridExtra)

.getFile <- function(path, pattern, .stop = T) {
    flist <- list.files(path = path, pattern = pattern, full.names = T)
    n <- length(flist)
    if (n != 1) {
        if (.stop) {
            stop(paste("Expected to find one file, but found", n, "from",
                       paste(flist, collapse = ", "), "instead"))
        } else {
            return(NA)
        }
    }
    return(flist[[1]])
}
```

\begin{comment}
Page 1 is a overview of the sample analysis
\end{comment}

`r if(!params$hasAnnot) {"\\begin{comment}"}`

Summary {data-orientation=rows}
=======================================================================

Row
-----------------------------------------------------------------------

### Sequence lengths without outlier removal

```{r, eval=params$hasAnnot, fig.width=params$width, fig.height=params$height}
fname <- .getFile(path = file.path(params$rootDir, "annot"),
                  pattern = ".*_all_clones_len_dist\\.Rdata")
load(fname)
if (params$interactive) {
    ggplotly(plot)
} else {
    plot
}
```

Row
-----------------------------------------------------------------------

### Sequence lengths __with__ outlier removal

```{r, eval=params$hasAnnot, fig.width=params$width, fig.height=params$height}
fname <- .getFile(path = file.path(params$rootDir, "annot"),
                  pattern = ".*_all_clones_len_dist_no_outliers\\.Rdata")
load(fname)
if (params$interactive) {
    ggplotly(plot)
} else {
    plot
}
```

`r if(!params$hasAnnot) {"\\end{comment}"}`

\begin{comment}
Page 2 is abundance analysis
\end{comment}

`r if(!params$hasAbun) {"\\begin{comment}"}`

Abundance {data-orientation=rows}
=======================================================================


Row
-----------------------------------------------------------------------

### V family 

```{r, eval=params$hasAbun, fig.width=params$width, fig.height=params$height}
fname <- .getFile(path = file.path(params$rootDir, "abundance"),
                  pattern = ".*igv_dist_family_level\\.Rdata")
load(fname)
if (params$interactive) {
    ggplotly(plot)
} else {
    plot
}
```

### V gene

```{r, eval=params$hasAbun, fig.width=params$height, fig.height=params$width}
fname <- .getFile(path = file.path(params$rootDir, "abundance"),
                  pattern = ".*igv_dist_gene_level\\.Rdata")
load(fname)
# TODO: ggplotly is buggy with horizontal plot in tabsets
if (params$interactive) {
    plot
    #ggplotly(plot)
} else {
    plot
}
```

> Only the top 15 from each sample are shown

`r if(!params$inclD) {"\\begin{comment}"}`

Row
-----------------------------------------------------------------------

### D family

```{r, eval=(params$inclD && params$hasAbun), fig.width=params$width, fig.height=params$height}
fname <- .getFile(path = file.path(params$rootDir, "abundance"),
                  pattern = ".*igd_dist_family_level\\.Rdata")
load(fname)
if (params$interactive) {
    ggplotly(plot)
} else {
    plot
}
```

### D gene

```{r, eval=(params$inclD && params$hasAbun), fig.width=params$height, fig.height=params$width}
fname <- .getFile(path = file.path(params$rootDir, "abundance"),
                  pattern = ".*igd_dist_gene_level\\.Rdata")
load(fname)
# TODO: ggplotly is buggy with horizontal plot in tabsets
if (params$interactive) {
    plot
    #ggplotly(plot)
} else {
    plot
}
```

> Only the top 15 from each sample are shown

`r if(!params$inclD) {"\\end{comment}"}`


Row
-----------------------------------------------------------------------

### J family

```{r, eval=params$hasAbun, fig.width=params$width, fig.height=params$height}
fname <- .getFile(path = file.path(params$rootDir, "abundance"),
                  pattern = ".*igj_dist_family_level\\.Rdata")
load(fname)
if (params$interactive) {
    ggplotly(plot)
} else {
    plot
}
```

### J variant

```{r, eval=params$hasAbun, fig.width=params$height, fig.height=params$width}
fname <- .getFile(path = file.path(params$rootDir, "abundance"),
                  pattern = ".*igj_dist_variant_level\\.Rdata")
load(fname)
# TODO: ggplotly is buggy with horizontal plot in tabsets
if (params$interactive) {
    plot
    #ggplotly(plot)
} else {
    plot
}
```

> Only the top 15 from each sample are shown

`r if(!params$single) {"\\begin{comment}"}`

Row
-----------------------------------------------------------------------

### V-J association

```{r, eval=(params$single && params$hasAbun), fig.width=params$width, fig.height=params$height}
fname <- .getFile(path = file.path(params$rootDir, "abundance"), pattern = ".*_vjassoc\\.png")
img <- readPNG(fname)
grid.raster(img)
```
> Plot associations between V genes and J genes on the family level

`r if(!params$single) {"\\end{comment}"}`



Row {.tabset .tabset-fade}
-----------------------------------------------------------------------

### Bitscore vs Alignment Length
```{r, eval=params$hasAbun, fig.width=params$width, fig.height=params$height}
fname <- .getFile(path = file.path(params$rootDir, "abundance"),
                  pattern = ".*_igv_align_quality_bitscore_hm\\.png")
img <- readPNG(fname)
grid.raster(img)
```

### Gaps vs Alignment Length
```{r, eval=params$hasAbun, fig.width=params$width, fig.height=params$height}
fname <- .getFile(path = file.path(params$rootDir, "abundance"),
                  pattern = ".*_igv_align_quality_gaps_hm\\.png")
img <- readPNG(fname)
grid.raster(img)
```

### Identity vs Alignment Length
```{r, eval=params$hasAbun, fig.width=params$width, fig.height=params$height}
fname <- .getFile(path = file.path(params$rootDir, "abundance"),
                  pattern = ".*_igv_align_quality_identity_hm\\.png")
img <- readPNG(fname)
grid.raster(img)
```

### Mismatches vs Alignment Length
```{r, eval=params$hasAbun, fig.width=params$width, fig.height=params$height}
fname <- .getFile(path = file.path(params$rootDir, "abundance"),
                  pattern = ".*_igv_align_quality_mismatches_hm\\.png")
img <- readPNG(fname)
grid.raster(img)
```

### Subject Start vs Query Start
```{r, eval=params$hasAbun, fig.width=params$width, fig.height=params$height}
fname <- .getFile(path = file.path(params$rootDir, "abundance"),
                  pattern = ".*_igv_align_quality_start_hm\\.png")
img <- readPNG(fname)
grid.raster(img)
```

Row
-----------------------------------------------------------------------

### Indels (Gaps)
```{r, eval=params$hasAbun, fig.width=params$width, fig.height=params$height}
fname <- .getFile(path = file.path(params$rootDir, "abundance"),
                  pattern = ".*_igv_gaps_dist\\.Rdata")
load(fname)
if (params$interactive) {
    ggplotly(plot)
} else {
    plot
}
```

> Plot of insertions and deletions found within the __V gene__

### Mismatches
```{r, eval=params$hasAbun, fig.width=params$width, fig.height=params$height}
fname <- .getFile(path = file.path(params$rootDir, "abundance"),
                  pattern = ".*_igv_mismatches_dist\\.Rdata")
load(fname)
if (params$interactive) {
    ggplotly(plot)
} else {
    plot
}
```

> Plot of nucleotide base mismatches found within the __V gene__

`r if(!params$hasAbun) {"\\end{comment}"}`


`r if(!params$hasProd) {"\\start{comment}"}`

Productivity {data-orientation=rows}
=======================================================================

Row
-----------------------------------------------------------------------

### Productivity summary

```{r, eval=params$hasProd, fig.width=params$width, fig.height=params$height}
fname <- .getFile(path = file.path(params$rootDir, "productivity"),
                  pattern = ".*_productivity\\.Rdata")
load(fname)
if (params$interactive) {
    # TODO: buggy facet plot
    plot
    # ggplotly(plot)
} else {
    plot
}
```

> Productivity distribution of sample based on stop codons and frameshifts

### Frameshift distribution

```{r, eval=params$hasProd, fig.width=params$width, fig.height=params$height}
fname <- .getFile(path = file.path(params$rootDir, "productivity"),
                  pattern = ".*_crossed_vjframe_dist\\.Rdata")

load(fname)
if (params$interactive) {
    ggplotly(plot)
} else {
    plot
}
```

> Distribution of in-frame and out-of-frame sequences


Row
-----------------------------------------------------------------------

### Mismatches, Gaps

Diagrams below, from left to right:


Distribution of __mismatches__, __gaps__ and __gaps__ of __out-of-frame__ sequences
within FR/CDR region and V, (D), J germlines.


> Gaps of __out-of-frame__ sequences are only shown for FR/CDR regions

Row {.tabset .tabset-fade}
-----------------------------------------------------------------------

### CDR1
```{r, eval=params$hasProd, fig.width=params$width*3, fig.height=params$height}
fname <- .getFile(path = file.path(params$rootDir, "productivity"),
                  pattern = ".*_cdr1_mismatches_dist\\.Rdata")
# take the titles away if it's interactive because subplots will only use the first
load(fname)
if (params$interactive) {
    mismatch <- ggplotly(plot + labs(title = ""))
} else {
    mismatch <- plot
}

fname <- .getFile(path = file.path(params$rootDir, "productivity"),
                  pattern = ".*_cdr1_gaps_dist\\.Rdata")
load(fname)
if (params$interactive) {
    gaps <- ggplotly(plot + labs(title = ""))
} else {
    gaps <- plot
}

fname <- .getFile(path = file.path(params$rootDir, "productivity"),
                  pattern = ".*_cdr1_gaps_dist_out_of_frame\\.Rdata")
load(fname)
if (params$interactive) {
    gapsOut <- ggplotly(plot + labs(title = ""))
} else {
    gapsOut <- plot
}

if (params$interactive) {
    subplot(mismatch, gaps, gapsOut, titleX = TRUE, titleY = TRUE)
} else {
    grid.arrange(mismatch, gaps, gapsOut, nrow = 1)
}
```

> Distribution of mismatches, gaps and gaps of __out-of-frame__ sequences in CDR1

### CDR2
```{r, eval=params$hasProd, fig.width=params$width*3, fig.height=params$height}
fname <- .getFile(path = file.path(params$rootDir, "productivity"),
                  pattern = ".*_cdr2_mismatches_dist\\.Rdata")
# take the titles away if it's interactive because subplots will only use the first
load(fname)
if (params$interactive) {
    mismatch <- ggplotly(plot + labs(title = ""))
} else {
    mismatch <- plot
}

fname <- .getFile(path = file.path(params$rootDir, "productivity"),
                  pattern = ".*_cdr2_gaps_dist\\.Rdata")
load(fname)
if (params$interactive) {
    gaps <- ggplotly(plot + labs(title = ""))
} else {
    gaps <- plot
}

fname <- .getFile(path = file.path(params$rootDir, "productivity"),
                  pattern = ".*_cdr2_gaps_dist_out_of_frame\\.Rdata")
load(fname)
if (params$interactive) {
    gapsOut <- ggplotly(plot + labs(title = ""))
} else {
    gapsOut <- plot
}

if (params$interactive) {
    subplot(mismatch, gaps, gapsOut, titleX = TRUE, titleY = TRUE)
} else {
    grid.arrange(mismatch, gaps, gapsOut, nrow = 1)
}
```

> Distribution of mismatches, gaps and gaps of __out-of-frame__ sequences in CDR2


### CDR3
```{r, eval=params$hasProd, fig.width=params$width*3, fig.height=params$height}
fname <- .getFile(path = file.path(params$rootDir, "productivity"),
                  pattern = ".*_cdr3_mismatches_dist\\.Rdata")
# take the titles away if it's interactive because subplots will only use the first
load(fname)
if (params$interactive) {
    mismatch <- ggplotly(plot + labs(title = ""))
} else {
    mismatch <- plot
}

fname <- .getFile(path = file.path(params$rootDir, "productivity"),
                  pattern = ".*_cdr3_gaps_dist\\.Rdata")
load(fname)
if (params$interactive) {
    gaps <- ggplotly(plot + labs(title = ""))
} else {
    gaps <- plot
}

fname <- .getFile(path = file.path(params$rootDir, "productivity"),
                  pattern = ".*_cdr3_gaps_dist_out_of_frame\\.Rdata")
load(fname)
if (params$interactive) {
    gapsOut <- ggplotly(plot + labs(title = ""))
} else {
    gapsOut <- plot
}

if (params$interactive) {
    subplot(mismatch, gaps, gapsOut, titleX = TRUE, titleY = TRUE)
} else {
    grid.arrange(mismatch, gaps, gapsOut, nrow = 1)
}
```

> Distribution of mismatches, gaps and gaps of __out-of-frame__ sequences in CDR3

### FR1
```{r, eval=params$hasProd, fig.width=params$width*3, fig.height=params$height}
fname <- .getFile(path = file.path(params$rootDir, "productivity"),
                  pattern = ".*_fr1_mismatches_dist\\.Rdata")
# take the titles away if it's interactive because subplots will only use the first
load(fname)
if (params$interactive) {
    mismatch <- ggplotly(plot + labs(title = ""))
} else {
    mismatch <- plot
}

fname <- .getFile(path = file.path(params$rootDir, "productivity"),
                  pattern = ".*_fr1_gaps_dist\\.Rdata")
load(fname)
if (params$interactive) {
    gaps <- ggplotly(plot + labs(title = ""))
} else {
    gaps <- plot
}

fname <- .getFile(path = file.path(params$rootDir, "productivity"),
                  pattern = ".*_fr1_gaps_dist_out_of_frame\\.Rdata")
load(fname)
if (params$interactive) {
    gapsOut <- ggplotly(plot + labs(title = ""))
} else {
    gapsOut <- plot
}

if (params$interactive) {
    subplot(mismatch, gaps, gapsOut, titleX = TRUE, titleY = TRUE)
} else {
    grid.arrange(mismatch, gaps, gapsOut, nrow = 1)
}
```

> Distribution of mismatches, gaps and gaps of __out-of-frame__ sequences in FR1

### FR2
```{r, eval=params$hasProd, fig.width=params$width*3, fig.height=params$height}
fname <- .getFile(path = file.path(params$rootDir, "productivity"),
                  pattern = ".*_fr2_mismatches_dist\\.Rdata")
# take the titles away if it's interactive because subplots will only use the first
load(fname)
if (params$interactive) {
    mismatch <- ggplotly(plot + labs(title = ""))
} else {
    mismatch <- plot
}

fname <- .getFile(path = file.path(params$rootDir, "productivity"),
                  pattern = ".*_fr2_gaps_dist\\.Rdata")
load(fname)
if (params$interactive) {
    gaps <- ggplotly(plot + labs(title = ""))
} else {
    gaps <- plot
}

fname <- .getFile(path = file.path(params$rootDir, "productivity"),
                  pattern = ".*_fr2_gaps_dist_out_of_frame\\.Rdata")
load(fname)
if (params$interactive) {
    gapsOut <- ggplotly(plot + labs(title = ""))
} else {
    gapsOut <- plot
}

if (params$interactive) {
    subplot(mismatch, gaps, gapsOut, titleX = TRUE, titleY = TRUE)
} else {
    grid.arrange(mismatch, gaps, gapsOut, nrow = 1)
}
```

> Distribution of mismatches, gaps and gaps of __out-of-frame__ sequences in FR2

### FR3
```{r, eval=params$hasProd, fig.width=params$width*3, fig.height=params$height}
fname <- .getFile(path = file.path(params$rootDir, "productivity"),
                  pattern = ".*_fr3_mismatches_dist\\.Rdata")
# take the titles away if it's interactive because subplots will only use the first
load(fname)
if (params$interactive) {
    mismatch <- ggplotly(plot + labs(title = ""))
} else {
    mismatch <- plot
}

fname <- .getFile(path = file.path(params$rootDir, "productivity"),
                  pattern = ".*_fr3_gaps_dist\\.Rdata")
load(fname)
if (params$interactive) {
    gaps <- ggplotly(plot + labs(title = ""))
} else {
    gaps <- plot
}

fname <- .getFile(path = file.path(params$rootDir, "productivity"),
                  pattern = ".*_fr3_gaps_dist_out_of_frame\\.Rdata")
load(fname)
if (params$interactive) {
    gapsOut <- ggplotly(plot + labs(title = ""))
} else {
    gapsOut <- plot
}

if (params$interactive) {
    subplot(mismatch, gaps, gapsOut, titleX = TRUE, titleY = TRUE)
} else {
    grid.arrange(mismatch, gaps, gapsOut, nrow = 1)
}
```

> Distribution of mismatches, gaps and gaps of __out-of-frame__ sequences in FR3

### IGV
```{r, eval=params$hasProd, fig.width=params$width*2, fig.height=params$height}
fname <- .getFile(path = file.path(params$rootDir, "productivity"),
                  pattern = ".*_igv_mismatches_dist\\.Rdata")
# take the titles away if it's interactive because subplots will only use the first
load(fname)
if (params$interactive) {
    mismatch <- ggplotly(plot + labs(title = ""))
} else {
    mismatch <- plot
}

fname <- .getFile(path = file.path(params$rootDir, "productivity"),
                  pattern = ".*_igv_gaps_dist\\.Rdata")
load(fname)
if (params$interactive) {
    gaps <- ggplotly(plot + labs(title = ""))
} else {
    gaps <- plot
}

if (params$interactive) {
    subplot(mismatch, gaps, titleX = TRUE, titleY = TRUE)
} else {
    grid.arrange(mismatch, gaps, nrow = 1)
}
```

> Mismatches and gaps distribution in V gene

`r if(!params$inclD) {"\\start{comment}"}`
### IGD
```{r, eval=(params$hasProd && params$inclD), fig.width=params$width*2, fig.height=params$height}
fname <- .getFile(path = file.path(params$rootDir, "productivity"),
                  pattern = ".*_igd_mismatches_dist\\.Rdata")
# take the titles away if it's interactive because subplots will only use the first
load(fname)
if (params$interactive) {
    mismatch <- ggplotly(plot + labs(title = ""))
} else {
    mismatch <- plot
}

fname <- .getFile(path = file.path(params$rootDir, "productivity"),
                  pattern = ".*_igd_gaps_dist\\.Rdata")
load(fname)
if (params$interactive) {
    gaps <- ggplotly(plot + labs(title = ""))
} else {
    gaps <- plot
}

if (params$interactive) {
    subplot(mismatch, gaps, titleX = TRUE, titleY = TRUE)
} else {
    grid.arrange(mismatch, gaps, nrow = 1)
}
```

> Mismatches and gaps distribution in D gene

`r if(!params$inclD) {"\\end{comment}"}`

### IGJ
```{r, eval=params$hasProd, fig.width=params$width*2, fig.height=params$height}
fname <- .getFile(path = file.path(params$rootDir, "productivity"),
                  pattern = ".*_igj_mismatches_dist\\.Rdata")
# take the titles away if it's interactive because subplots will only use the first
load(fname)
if (params$interactive) {
    mismatch <- ggplotly(plot + labs(title = ""))
} else {
    mismatch <- plot
}

fname <- .getFile(path = file.path(params$rootDir, "productivity"),
                  pattern = ".*_igj_gaps_dist\\.Rdata")
load(fname)
if (params$interactive) {
    gaps <- ggplotly(plot + labs(title = ""))
} else {
    gaps <- plot
}

if (params$interactive) {
    subplot(mismatch, gaps, titleX = TRUE, titleY = TRUE)
} else {
    grid.arrange(mismatch, gaps, nrow = 1)
}
```

`r if(!params$hasProd) {"\\end{comment}"}`

> Mismatches and gaps distribution in J gene


Row {.tabset .tabset-fade}
-----------------------------------------------------------------------

### IGV productive distribution

```{r, eval=params$hasProd, fig.width=params$width, fig.height=params$height}
fname <- .getFile(path = file.path(params$rootDir, "productivity"),
                  pattern = ".*_igv_dist_productive\\.Rdata")

load(fname)
if (params$interactive) {
    ggplotly(plot)
} else {
    plot
}
```

> Distribution of productive sequences from IGV family

### IGV inframe, unproductive distribution

```{r, eval=params$hasProd, fig.width=params$width, fig.height=params$height}
fname <- .getFile(path = file.path(params$rootDir, "productivity"),
                  pattern = ".*_igv_dist_inframe_unproductive\\.Rdata")

load(fname)
if (params$interactive) {
    ggplotly(plot)
} else {
    plot
}
```

> Distribution of in-frame but unproductive sequences from IGV family

### IGV out-of-frame distribution

```{r, eval=params$hasProd, fig.width=params$width, fig.height=params$height}
fname <- .getFile(path = file.path(params$rootDir, "productivity"),
                  pattern = ".*_igv_dist_out_of_frame\\.Rdata")

load(fname)
if (params$interactive) {
    ggplotly(plot)
} else {
    plot
}
```

> Distribution of out-of-frame sequences from IGV family


Row {.tabset .tabset-fade}
-----------------------------------------------------------------------

### Inframe

```{r, eval=params$hasProd, fig.width=params$width, fig.height=params$height}
fname <- .getFile(path = file.path(params$rootDir, "productivity"),
                  pattern = ".*_stopcodon_region_inframe\\.Rdata", .stop = F)
if (is.na(fname)) {
    cat(paste("Plot is not available for this sample, probably because there's",
                "no such distribution present in this sample"))
} else {
    load(fname)
    if (params$interactive) {
        ggplotly(plot)
    } else {
        plot
    }
}

```

> Distribution of stop codons in CDR/FR regions from in-frame sequences

### Out-of-frame
```{r, eval=params$hasProd, fig.width=params$width, fig.height=params$height}
fname <- .getFile(path = file.path(params$rootDir, "productivity"),
                  pattern = ".*_stopcodon_region_outframe\\.Rdata", .stop = F)
if (is.na(fname)) {
    cat(paste("Plot is not available for this sample, probably because there's",
                "no such distribution present in this sample"))
} else {
    load(fname)
    if (params$interactive) {
        ggplotly(plot)
    } else {
        plot
    }
}
```

> Distribution of stop codons in CDR/FR regions from out-of-frame sequences


\begin{comment}
=======================================================================
-----------------------------------------------------------------------
\end{comment}
