---
params:
    - slow: FALSE
title: "Introduction to abseqR"
author: "Jia Hong Fong"
package: "abseqR"
output:
    BiocStyle::html_document
vignette: >
    %\VignetteIndexEntry{Introduction to abseqR}
    %\VignetteEngine{knitr::rmarkdown}
    %\VignetteEncoding{UTF-8}
bibliography: bibliography.bib
---

```{r setup, include=FALSE}
library(abseqR)
```

```{r environment_setup, include=FALSE, eval=params$slow}
# loads the necessary variables for this Rmd
sandboxDirectory <- "~/sandbox"
if (!file.exists(sandboxDirectory)) {
    file.create(sandboxDirectory)
}
toyData <- system.file("extdata", "abseqR_example", package = "abseqR")
file.copy(toyData, sandboxDirectory, recursive = TRUE)
samples <- suppressWarnings(abseqReport(file.path(sandboxDirectory,
                                                  "abseqR_example"), report = 0
                                        )
                            )

```

```{r abseqR_vignette_helper_functions, include=FALSE}
sabseqR <- function() { return(Biocpkg("abseqR")); }
#sabseqR <- function() { return("`abseqR`"); }
sabseqPy <- function() { return("`abseqPy`"); }
sabseq <- function() { return("`AbSeq`"); }
```

# Introduction

A plethora of high throughput sequencing (HTS) analysis pipelines are
available as open source tools to analyze and validate the quality of Rep-seq [^Repseq] datasets.
[OmicTools](https://omictools.com/search?q=repertoire)
provides a summary of repertoire sequencing tools that implements different
techniques and algorithms in analyzing and visualizing datasets from B-cell
receptors (BCR) and T-cell receptors (TCR). However, high throughput antibody
sequencing analysis pipelines are scarce.

`r sabseq()` is a family of tools
implemented in `R`(visualizations and statistics) and `Python`(backend algorithm)
meant to be used together to provide insight to antibody repertoires.
Under the umbrella of `r sabseq()` is `r sabseqPy()` which processes paired-end
or single ended _FASTA/FASTQ_ files from Illumina MiSeq [@Illumina] into
csv files and HDF [@hdf] dataframes. The other is `r sabseqR()`, an R package that
visualizes the output of `r sabseqPy()` and presents the plots in a self-contained
HTML report. Furthermore, `r sabseqR()` provides an additional feature to
explicitly compare multiple samples against each other.

`r sabseqR()` provides the following functionality in addtion to `r sabseqPy()`:

* Visualizations: The output from `r sabseqPy()` is summarized succintly into
  plots saved as `PNG` and `Rdata` files. The latter provides flexibility for
  users to easily customize the aesthetics of any plots generated by `r sabseqR()`
 
* Collated report: The plots generated by `r sabseqR()` are collated and presented
  in a self-contained HTML document for convenience and ease of sharing.
  
* Sample comparison: `r sabseqR()` overloads the `+` operator
  via the S4 classes `AbSeqCRep` and `AbSeqRep`. To compare
  different samples against each other, the `+` operator can be used to
  specify the desired combinations. The "sample comparison" includes additional
  plots, for example, sample similarity clustering, overlapping clonotypes and
  more.
  

# Citation

TODO


# Installation

`r sabseqR()` can be installed from [bioconductor.org](http://bioconductor.org/) or
its GitHub repository at https://github.com/malhamdoosh/abseqR.

## Bioconductor

Install via `biocLite`:
```r
source("http://bioconductor.org/biocLite.R")
biocLite("abseqR")
```

## GitHub

Install via `devtools`:
```r
library(devtools)
install_github("malhamdoosh/abseqR")
```

## System prerequisites

`r sabseqR()` requires [pandoc](http://pandoc.org/) version _1.19.2.1_ or higher
to render the HTML report. If pandoc can not be detected while executing `r sabseqR()`,
the HTML report will __not__ be generated. `r sabseqR()` is a cross-platform
library and will work on any major operating system [^windows_biocparallel].

## R package dependencies

`r sabseqR()` depends on several packages from [CRAN](https://cran.r-project.org/)
and [Bioconductor](https://bioconductor.org/):

* [RColorBrewer](https://cran.r-project.org/package=RColorBrewer) provides colour
schemes for maps and graphics generated in this package. To install it, type 
`install.packages("RColorBrewer")`

* [VennDiagram](https://cran.r-project.org/package=VennDiagram) provides a set of
functions to generate Venn diagrams. To install it, type `install.packages("VennDiagram")`

* [circlize](https://cran.r-project.org/package=circlize) is a visualization tool
used to summarize the distributions of associations between segments in this
package. To install it, type `install.packages("circlize")`

* [flexdashboard](https://cran.r-project.org/package=flexdashboard) is a package
that provides a template for RMarkdown that resembles a grid oriented dashboard.
This is used to generate the HTML report in a dashboard fashion. To
install it, type `install.packages("flexdashboard")`

* [ggplot2](https://cran.r-project.org/package=ggplot2) is an implementation of
the "Grammar of Graphics" in R. It is used extensively within this package to
generate plots. To install it, type `install.packages("ggplot2")`

* [ggcorrplot](https://cran.r-project.org/package=ggcorrplot) is used to visualize
a correlation matrix using `ggplot2`. To install it, type `install.packages("ggcorrplot")`

* [ggdendro](https://cran.r-project.org/package=ggdendro) provides a set of tools
for drawing dendrograms and tree plots using `ggplot2`. To install it, type
`install.packages("ggdendro")`

* [grid](https://cran.r-project.org/package=grid) is used to arrange plots.
It has been integrated into the base R package.

* [gridExtra](https://cran.r-project.org/package=gridExtra) provides functions
to work with "grid" graphics, used in this package to arrange grid-based plots
in the HTML report. To install it, type `install.packages("gridExtra")`

* [knitr](https://cran.r-project.org/package=knitr) provides the capability to
dynamically generate reports in R. To install it, type `install.packages("knitr")`

* [plotly](https://cran.r-project.org/package=plotly) is used to translate
`ggplot2` graphs to an interactive web-based version. To install it, type
`install.packages("plotly")`

* [plyr](https://cran.r-project.org/package=plyr) offers a set of tools used in
this package to apply operations on subsets of data in manageable pieces. To
install it, type `install.packages("plyr")`

* [png](https://cran.r-project.org/package=png) is used to read and display PNG images.
To install it, type `install.packages("png")`

* [reshape2](https://cran.r-project.org/package=reshape2) allows this package to
restructure and aggregate dataframes. To install it, type `install.packages("reshape2")`

* [rmarkdown](https://cran.r-project.org/package=rmarkdown) converts R Markdown
documents into a variety of formats. To install it, type `install.packages("rmarkdown")`

* [vegan](https://cran.r-project.org/package=vegan) [@vegan] provides a suite of functions
to calculate diversity and distance statistics between repertoires. To install it,
type `install.packages("vegan")`

* [BiocParallel](https://bioconductor.org/packages/release/bioc/html/BiocParallel.html) is
a package from [Bioconductor](https://bioconductor.org) used to parallelize
operations in this package. To install it, type
```r
## try http:// if https:// URLs are not supported
        source("https://bioconductor.org/biocLite.R")
        biocLite("BiocParallel")
```

# Quick start

To leverage _all_ the functionalities provided by `r sabseqR()`, the main
functions to note are `abseqR::abseqReport`, `abseqR::report`, and `+`.
This section will use a toy example to walk through the use cases of each function.

## Dataset

The toy example mentioned previously comes installed with `r sabseqR()`.
For the sake of brevity, the samples are described under the
[miscellaneous](#toy-data) section. Briefly, the dataset contains 5 samples generated
from _in silico_ simulation, each named PCR1, PCR2, ... and so on.

### Obtaining the toy dataset {#obtain-dataset}

To obtain the dataset, the following R code demonstrates the steps needed to copy it
to a _sandbox directory_ for play-testing:

```{r abseqR_copy_data_files, eval=FALSE}
# substitute with any directory that you have read/write access to
sandboxDirectory <- "~/sandbox"
# path to provided toy data (comes installed with abseqR)
toyData <- system.file("extdata", "abseqR_example", package = "abseqR")
# copy the provided toy data to sandboxDirectory
file.copy(toyData, sandboxDirectory, recursive = TRUE)
```

### Exploring the toy dataset

The contents of `abseqR_example` should be familar to users of `r sabseqPy()`.
```{r abseqR_data_contents}
# toyDataPath now holds the path to a local copy of the toy data
toyDataPath <- file.path(sandboxDirectory, "abseqR_example")
list.files(path = toyDataPath)
```

Peeking into the `auxiliary` folder, the names of the samples are:
```{r abseqR_sample_names}
paste(list.files(path = file.path(toyDataPath, "auxiliary")), collapse = ", ")
```


## Analyzing the datasets

### Basic analysis with `abseqReport`

After [obtaining the dataset](#obtain-dataset), visualizing the results of
`r sabseqPy()` is done using the `abseqReport` function:

```{r abseqR_quick_run, eval=FALSE}
# visualize the datasets, then compare PCR1, PCR2, PCR3, PCR4, and PCR5
samples <-  abseqReport(toyDataPath, compare = c("PCR1, PCR2, PCR3, PCR4, PCR5"))
# ignore the warning message:
# "Sample output directory <path> is different from provided path
#   <path> assuming directory was moved"
# This warning message tells us that the directory has
# been moved (we copied it elsewhere)
```
Here, `abseqReport` was told to also compare samples
`PCR1`, `PCR2`, `PCR3`, `PCR4`, and `PCR5` in addition to analyzing the samples
within `toyDataPath`. If sample comparison was not required, then the command can be
simplified to `samples <- abseqReport(toyDataPath)`.


`abseqReport` returns a named list of objects[^abseqRep], each object encapsulates
information about the samples within `toyDataPath`.

```{r abseqR_named_list}
# PCR1 - PCR5
stopifnot(length(samples) == 5)
# display the samples found within "toyDataPath"
names(samples)
```

Now is a good time to browse through ``r file.path(toyDataPath, "report")``, where
`index.html` should be present.

```{r}
list.files(path = file.path(toyDataPath, "report"))
```

* `index.html` is a document which summarizes the analysis and provides a convenient way
of navigating to multiple samples. For example, within this file, there are links to
the reports generated for `PCR1`, `PCR2`,`PCR3`,`PCR4`, `PCR5`, and `PCR1 vs ... vs PCR5`.

* `html_files` contains individual HTML files that are _not_ meant to be accessed by the user,
but rather through `index.html`. The files within this directory are standalone HTML documents,
but the landing page `index.html` (is **not** because it) contains relative links to the aforementioned HTML documents,
and should be shared together with the `html_files` folder, if required.

In conclusion, the folder of interest is `report`, and the file of interest is
`index.html`. The individual sample reports (HTML files) in `html_files` can
be shared as-is, but it is recommended to share the `report` folder in its entirety.


### Customizing sample comparison with `report`

Previously, the return value of `abseqReport` was stored in a variable
named `samples`. This allows for further customizations in sample comparison.

<!--
1. describe what the objects in samples can now be used for (+ operator)
    1.1 use the report() function
2. describe a quick way of getting the objects without ever plotting (report = 0)
-->


# Advanced Examples

<!--
1. describe other settings of report = 0, 1, 2, 3
2. describe how to load and compare different samples from different directories
-->

<!-- 1. multi-output directories -->
<!-- 2. BPPARAM usage -->

# Interpreting the analysis reports




# Miscellaneous

## Dataset {#toy-data}

The toy dataset used in the above examples
was obtained from a combination of synthetic sample datasets generated
using [MiXCR](https://github.com/milaboratory/mixcr)'s program 
[here](http://files.milaboratory.com/mixcr/paper/mixcr-test-1.2-SNAPSHOT.jar). 
Firstly, three distinct samples were generated, each simulated with the following parameters
in `MiXCR`:
```
* reads   = 10000, 10000, 10000
* clones  =  5000,  5000,  2000
* seed    =  4228,  2428,  2842
* conf    = MiSeq-300-300      (for all 3)
* loci    = IGH                (for all 3)
* species = hsa                (for all 3)
```

Following that, to introduce overlapping sequences between the samples, a script was written
to randomly mix and amplify sequences from these samples, resulting in a final
repertoire of 5 samples, named PCR1, PCR2, and so on.

Finally, these 5 samples were analyzed by `r sabseqPy()`. The command used to
analyze these samples are as follows:

```{bash abseqR_toy_data_abseqPy_run, eval=FALSE}
abseq -y toy.yml
```

where the contents of `toy.yml` is:

```yaml
# toy.yml
defaults:
    bitscore: 300
    sstart: 1-3
    alignlen: 250
    outdir: abseqR_example
    task: all
    threads: 1
---
file1: PCR1.fasta
name: PCR1
---
file1: PCR2.fasta
name: PCR2
---
file1: PCR3.fasta
name: PCR3
---
file1: PCR4.fasta
name: PCR4
---
file1: PCR5.fasta
name: PCR5
```

These 5 samples form the final toy dataset described in this vignette.

# Session Info

```{r abseqR_session_info, echo=FALSE}
sessionInfo()
```

[^Repseq]: Repertoire sequencing
[^windows_biocparallel]: The performace offered by
[BiocParallel](https://bioconductor.org/packages/release/bioc/html/BiocParallel.html) may differ across OSes
[^abseqRep]: The "objects" returned are `AbSeqRep` S4 class instances.

# References
<!-- automatically inserted -->
